src/Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design/Internal/CodeGenerator.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using Microsoft.AspNetCore.Mvc.Core.ApplicationParts;
+using Microsoft.AspNetCore.Mvc.Razor.Internal;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Precompilation.Design.Internal
+{
+    public class CodeGenerator
+    {
+        public CodeGenerator(
+            CSharpCompiler compiler,
+            CSharpCompilation compilation)
+        {
+            Compiler = compiler;
+            Compilation = compilation;
+        }
+
+        public CSharpCompiler Compiler { get; }
+
+        public CSharpCompilation Compilation { get; private set; }
+
+        public void AddViewFactory(List<ViewCompilationInfo> result)
+        {
+            var precompiledViewsArray = new StringBuilder();
+            foreach (var item in result)
+            {
+                precompiledViewsArray.AppendLine(
+                    $"new {typeof(PrecompiledViewInfo).FullName}(@\"{item.RelativeFileInfo.RelativePath}\", typeof({item.TypeName})),");
+            }
+
+            var factoryContent =
+            $@"
+namespace {AssemblyPart.ViewFactoryNamespace}
+{{
+  public class {AssemblyPart.ViewFactoryTypeName} : {typeof(PrecompiledViews).FullName}
+  {{
+    public {AssemblyPart.ViewFactoryTypeName}() : base(new[]
+    {{
+        {precompiledViewsArray}
+    }})
+    {{
+    }}
+  }}
+}}";
+            var syntaxTree = Compiler.CreateSyntaxTree(SourceText.From(factoryContent));
+            Compilation = Compilation.AddSyntaxTrees(syntaxTree);
+        }
+
+        public void AddAssemblyMetadata(AssemblyName applicationAssemblyName, string keyFilePath)
```

Comment:
Looks like we're only ever passing in null here from call sites. Is that intentional?